MNT: add `astropy.system_info` report helper function

neutrinoceros opened the pull request:
### Description
This has been proposed a couple times in passing, so here's an implementation of this idea that's immediately useful (not just when every user can be assumed to have a sufficiently recent version of astropy that includes the new function).

this outputs a report looking like
```
platform
--------
platform.platform() = 'macOS-14.4.1-arm64-arm-64bit'
platform.version() = 'Darwin Kernel Version 23.4.0: Fri Mar 15 00:12:49 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6020'
platform.python_version() = '3.12.2'

packages
--------
astropy              7.0.0.dev41+gdcf60561fc
numpy                1.26.4
scipy                1.13.0
matplotlib           --
pandas               --
pyerfa               2.0.1.4
``` 

How to test this:
```
curl --silent https://raw.githubusercontent.com/neutrinoceros/astropy/mnt/report_info_script/astropy/utils/report_info.py | python
```

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

astrofrog commented:
I like this idea in general, but I am wondering if there is any benefit to including the list of packages here versus just asking users to give the output of pip freeze?

astrofrog commented:
This might also make more sense to put in https://github.com/astropy/astropy-tools - and it doesn't have to be core package specific, but could be used for coordinated packages too.

neutrinoceros commented:
> I am wondering if there is any benefit to including the list of packages here versus just asking users to give the output of pip freeze?

I can think of one: not everybody uses pip (conda and uv come to mind)

neutrinoceros commented:
> This might also make more sense to put in [astropy/astropy-tools](https://github.com/astropy/astropy-tools) - and it doesn't have to be core package specific, but could be used for coordinated packages too.

I agree. However, I'm not sure how much additional complexity would be required to make it re-usable (the list of "relevant" packages would vary from one repo to the other)

neutrinoceros commented:
Another reason why *not* to make this a shared resource is that I would like to add information about the build environment, eventually, but that's so project specific that I don't think such a section could be made portable, could it ?

pllim commented:
`curl` isn't Windows friendly, so I don't think this would fly as it is even though it is a good idea.

neutrinoceros commented:
Ah, thanks for pointing this out. I don't know off-hand how to change it but I definitely don't want to alienate such a common platform.

astrofrog commented:
Thinking about this a bit more, this seems to overlap somewhat with pytest-astropy-header - certainly what we are printing here is very similar to what is there. Can we avoid duplication somehow? We could for instance ask users to install pytest and do:

```
pytest --pyargs astropy --astropy-header
```

which has output that includes:

```
Platform: Linux-6.5.0-28-generic-x86_64-with-glibc2.35

Executable: /home/tom/python/dev/bin/python3.11

Full Python Version: 
3.11.9 (main, Apr  6 2024, 17:59:24) [GCC 11.4.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.25.2
Scipy: 1.11.2
Matplotlib: 3.7.3
h5py: 3.10.0
Pandas: 2.1.0
```

anyway not saying that is the solution, but there certainly is a lot of overlap, and various packages have already put in the work to determine what is relevant to show in this list.

neutrinoceros commented:
Ideally there should be as little friction as possible, and requiring pytest is a bit much in my opinion: users with broken env or not able to install python packages (for example if they use a server where they don't have install rights) might turn away the minute we ask them to do something that doesn't work out of the box.

The invoke you're proposing also *runs* a large part of the test suite, which is a lot more than what my script intends to do.
Do you know of a way to ask pytest to output *just* the header ?

All that said, it might indeed be a good idea to avoid duplication, but I'm thinking it should be the other way around (pytest header would rely on this script), if possible.

pllim commented:
I think there was a very old conversation about such thing but no one could agree on what is the one best way to pull it off. Every problem is a little different.

eerovaher commented:
In our bug report template we ask the users to list the versions of some of the key packages, but the less code they need to run for that the better. Relying on any external code (whether it be from `pytest-astropy-header` or somewhere else) is not helpful for this purpose. 

Having a single command that could be copy-pasted and executed in a shell would be good, but in practice it doesn't work on all platforms. Possible problems using `curl` on Windows were already mentioned, and independently from that Debian-based systems (e.g. Ubuntu) have `python3` instead of `python`. It looks like the best we could do is to require users to execute two Python statements: the first to import a function and the second to run it.

neutrinoceros commented:
> It looks like the best we could do is to require users to execute two Python statements: the first to import a function and the second to run it.

That's indeed how I intend it to be used long term, when it is safe to assume that every user runs a sufficiently recent version of astropy that it has the function, but it can't be how we advertise it from the get go. What if w *didn't* try to use it in the bug report template for now (or maybe just suggest it in passing) ?

eerovaher commented:
There's no LTS version anymore, so we no longer provide bugfixes for `astropy` versions older than 6 months. Furthermore, we recommend that users reporting a bug should verify that the bug is still present in the development version of `astropy`. That means that the transition period during which both new and old instructions should remain visible can be rather short.

neutrinoceros commented:
Ok let's go with this plan then.

neutrinoceros commented:
In order to avoid making it more confusing than it needs to, I streamlined the instructions into a `try/except` block. The script will only used if installed, so it can also serve as an indicator of where our users are at. And curl isn't needed any more, so it should behave the same on all platforms.

pllim commented on .github/ISSUE_TEMPLATE/bug_report.yaml:
```diff
@@ -40,25 +40,28 @@ body:
-            print("Matplotlib", matplotlib.__version__)
-        except ImportError:
-            print("Matplotlib not installed")
+            from astropy.utils.report_info import main; main()
```
Still not 100% sold on this. This import won't work until astropy 7.0 is released. Until then, we might just get confused users reporting that our bug report template is broken.

I agree with @astrofrog that https://github.com/astropy/astropy-tools is probably a better place for such scripts, but then the same problem about it not being readily installable.

neutrinoceros replied:
> This import won't work until astropy 7.0 is released.

except for reporters running the main branch, which I suspect wouldn't be a negligible fraction ?

> Until then, we might just get confused users reporting that our bug report template is broken.

Now *I*'m confused: why would anyone report this as broken if there's not visible exception being raised ? (this isn't a rhetorical question, I'm genuinely curious)

pllim replied:
Actually a lot of users report from stable release. Not everyone installs from `main` even when requested for various reasons (no time, no desire, stuck in old release, etc). Therefore for those users, `from astropy.utils.report_info import main` will not work until we release v7. Hope this clarifies the matter.

eerovaher replied:
But this is in a `try-except` block.

pllim replied:
Oh, oops... Must have missed it in the diff. Let me ponder again then when I get the chance. Thanks for your patience.

pllim replied:
Would it be more explicit if we now do an `astropy` version check instead of a blind try-except?

neutrinoceros replied:
More explicit yes, but maybe also more fragile, and I'm afraid it would come back biting us with users reporting the script is broken...

pllim replied:
Well, we would need to test this script in CI anyway. And even then if it is still broken, won't we want to know? Under what condition would the script not break in CI but broken when user runs it? 🤔

neutrinoceros replied:
Since the attribute `astropy.__version__` is generated via `setuptools_scm`, I don't feel confortable taking any bets on exactly what could go wrong, but I don't trust that CI exercises all the ways it can be broken, which should all be weird edge cases that no one ever reported yet.

neutrinoceros replied:
Additionally, I note that
1) the main function in the script has a doctest
2) the exact way it's used in the report template is *not* exercised yet (as far as I know). That's definitely doable if you think it's a good idea.

eerovaher commented on astropy/utils/report_info.py:
````diff
@@ -0,0 +1,98 @@
+r"""
+A helper script to automate system info in bug reports.
+
+This can be run in two fashions either by downloading the latest version of this
+script and running it on the fly
+```
+curl --silent https://raw.githubusercontent.com/astropy/astropy/main/astropy/utils/report_info.py | python
+```
````
I'd remove this entirely. It is not possible to provide a simple shell one-liner that will work on all the supported platforms.

eerovaher commented on astropy/utils/report_info.py:
```diff
@@ -0,0 +1,98 @@
+from itertools import chain
+
+
+def header(name):
+    return [name, "-" * len(name)]
```
Adding type annotations is not mandatory, but I recommend adding them nonetheless:
```suggestion
def header(name: str) -> tuple[str, str]:
    return name, len(name) * "-"
```
or perhaps
```suggestion
def header(name: str) -> str:
    return f"{name}\n{len(name) * '-'}"
```

neutrinoceros replied:
I'd like to keep the return type a list for two reasons:
- in tuples, position is significant (borrowing the "Cultural Difference between lists and tuples" from https://nedbatchelder.com/blog/201608/lists_vs_tuples.html)
- your second suggestion requires more changes

eerovaher replied:
I'd prefer to keep the output immutable unless there is good reason to make it mutable, but these functions should never be called outside this module, so it doesn't really matter.

eerovaher commented on astropy/utils/report_info.py:
```diff
@@ -0,0 +1,98 @@
+
+
+def report_platform():
+    lines = [
+        f"{platform.platform() = }",
+        f"{platform.version() = }",
+        f"{platform.python_version() = }",
+    ]
+
+    return lines
```
```suggestion
    return (
        f"{platform.platform() = }",
        f"{platform.version() = }",
        f"{platform.python_version() = }",
    )
```
or maybe
```suggestion
    return "\n".join(
        [
            f"{platform.platform() = }",
            f"{platform.version() = }",
            f"{platform.python_version() = }",
        ]
    )
```

eerovaher commented on astropy/utils/report_info.py:
```diff
@@ -0,0 +1,98 @@
+        "matplotlib",
+        "pandas",
+    ]
+    lines = []
+    for p in pkgs:
+        if find_spec(p) is None:
+            v = "--"
+        else:
+            v = version(p)
+        lines.append(f"{p:<20} {v}")
```
```suggestion
    lines = [f"{p:<20} {'--' if find_spec(p) is None else version(p)}" for p in pkgs]
```

eerovaher commented on astropy/utils/report_info.py:
```diff
@@ -0,0 +1,98 @@
+    pandas               ...
+    pyerfa               ...
+    """
+    report_lines = chain(
```
I'm not convinced importing `itertools` just to use `chain()` here once is worth it.

neutrinoceros replied:
I'm confused, what does it cost ? Last I checked `itertools` wasn't an expensive import, and it seems to be imported so often by other parts of the standard library that trying to avoid the overhead seemed futile.

eerovaher replied:
`chain()` is not an overly expensive function, but in this case it is very simple to avoid importing `itertools` entirely:
```suggestion
    report_lines = "\n".join([*header("platform"), *report_platform(), "", *header("packages"), *report_packages()])
```
or if you move the `header()` calls inside the `report_*` functions it would be as simple as
```suggestion
    report_lines = "\n".join([*report_platform(), "", *report_packages()])
```

neutrinoceros replied:
I've dropped itertools using unpacking. Thanks !

neutrinoceros commented on astropy/utils/report_info.py:
```diff
@@ -0,0 +1,98 @@
+    pyerfa               ...
+    """
+    report_lines = chain(
+        header("platform"),
```
separation of concerns, I guess, but I don't hold a strong opinion on this.

eerovaher commented on astropy/utils/report_info.py:
````diff
@@ -0,0 +1,84 @@
+
+This can be run as
+```
+python -c "from astropy.utils.report_info import main; main()"
````
This might not work on Debian-based systems (including Ubuntu) because on those Python is by default installed as `python3`, not `python`.

neutrinoceros replied:
So, what are you suggesting ? should I mention it in text or should I just present the 2-line python script as such ?

neutrinoceros replied:
in passing, I'm tempted to go with `python -m astropy.utils.report_info` instead (already implemented)

eerovaher commented:
I'd prefer if the function that prints the installed versions was named like either [`numpy.show_runtime()`](https://numpy.org/doc/stable//reference/generated/numpy.show_runtime.html) or possibly [`sunpy.system_info()`](https://docs.sunpy.org/en/stable/generated/api/sunpy.util.system_info.html).

pllim commented:
Thinking about this more, I am curious to see how other packages are rolling such things out. I have reached out to Scientific Python and will get back here after I have gathered some external data. Thank you for your continued patience!

pllim commented:
My conclusion from various feedback is that the packages rolled their own reporting because each has different needs. So the script is probably okay in this repo, but maybe ditch the CLI and rename `main()` to something more descriptive?

### scipy

`scipy.show_config()`

https://github.com/scipy/scipy/blob/1464ae12df417a8c8b6750d86fca4437d1c732c0/scipy/__config__.py.in#L122

https://github.com/scipy/scipy/blob/1464ae12df417a8c8b6750d86fca4437d1c732c0/scipy/__init__.py#L51

### numpy

`np.show_runtime()`

https://github.com/numpy/numpy/blob/aa0cc043d1fe51e61024842b21905e11a2a4cee6/numpy/lib/_utils_impl.py#L20

### Others

`pyvista.Report()`

`mne.sys_info()`

pllim commented:
`system_info` like the only sunpy uses (ref https://github.com/astropy/astropy/pull/16335#issuecomment-2087385647) seems reasonable.

pllim commented on astropy/utils/report_info.py:
Add an `__all__` in this module as well?

neutrinoceros commented:
> maybe ditch the CLI

I'm not sure what you mean here. Are you refering to the `curl` invoke (already ditched but may still appear in outdated threads), or the `if __name__ == "__main__":` block ? (in the latter case, may I ask why ? I don't see the hurt in keeping it around)

Thanks for reminding me of @eerovaher's suggestion to change the name of the function, I think I missed it entirely !

neutrinoceros commented:
I've renamed the module and its main function, added an `__all__`, and re-exposed the main function as `astropy.system_info` since having it at the package's top level seems to be the convention in the ecosystem.
The branch history should get squashed into a single commit with the same title as the PR. Happy to do this by hand if desired.

eerovaher commented on astropy/utils/system_info.py:
````diff
@@ -0,0 +1,87 @@
+
+This can be run as
+```
+python -m astropy.utils.report_info
````
```suggestion
python -m astropy.utils.system_info
```

eerovaher commented on astropy/utils/system_info.py:
```diff
@@ -0,0 +1,87 @@
+This script should not require anything else than the standard library.
+"""
+
+__all__ = ["system_info"]
```
Defining `__all__` makes this module look public, so there should be an explicit warning in the module docstring that the contents here are subject to change without notice.

eerovaher commented on astropy/utils/system_info.py:
```diff
@@ -0,0 +1,87 @@
+    """
+    Examples
+    --------
+    >>> main()  # doctest: +ELLIPSIS
```
```suggestion
    >>> import astropy
    >>> astropy.system_info()
```
I have a vague memory that we don't need `+ELLIPSIS` because it is turned on globally somewhere in our configuration, but do check if this is true.

eerovaher commented on astropy/utils/system_info.py:
```diff
@@ -0,0 +1,87 @@
+from importlib.util import find_spec
+
+
+def header(name: str) -> list[str]:
```
I'd prefer if all the functions except `system_info()` would be explicitly marked as private:
```suggestion
def _header(name: str) -> list[str]:
```

eerovaher commented on astropy/utils/system_info.py:
```diff
@@ -0,0 +1,88 @@
+    """
+    Examples
+    --------
+    >>> system_info()
```
This is good enough for doctests, but users would find it more helpful if the example would be reproducible in a freshly started Python session:
```suggestion
    >>> import astropy
    >>> astropy.system_info()
```

pllim commented on astropy/__init__.py:
```diff
@@ -234,3 +234,6 @@ def online_help(query):
 del varname, __module_type__
+
+
+from astropy.utils.system_info import system_info
```
Does this also need to be updated?

https://github.com/astropy/astropy/blob/e7a556fdeda076f72277b06a8ef6bf9acc29b8f4/astropy/__init__.py#L199

neutrinoceros replied:
you tell me. I have never seen this variable name before, I don't know what it does.

pllim replied:
Looks like @saimn renamed it from `__dir__` in https://github.com/astropy/astropy/pull/7617 . Do you remember why we need this, Simon?

saimn replied:
Yes, `module.__dir__` has a special meaning since Python 3.7: https://docs.python.org/3.12/reference/datamodel.html#customizing-module-attribute-access

neutrinoceros replied:
Oh, I see. I think my new function *does* belong in `__dir_inc__`, which lists everything that should *not* be dropped from the namespace, but since I added it at the end of the module, it's only included after namespace cleanup happens, so it dodges that. I think the cleanup ought to be the last thing to happen in the module, so I'll move my import and update this list, thanks !

pllim commented on astropy/utils/system_info.py:
```diff
@@ -0,0 +1,89 @@
+    return 0
+
+
+if __name__ == "__main__":
```
I still don't see why we need this part.

pllim replied:
@eerovaher , what do you think?

eerovaher replied:
I wouldn't have implemented this. The ability to run
```python
>>> import astropy
>>> astropy.system_info()
```
in Python should be good enough.

neutrinoceros replied:
Fair enough I'll drop it

pllim commented:
> in the latter case, may I ask why ? I don't see the hurt in keeping it around

Just one more thing we have to maintain but not advertised to be used in the template anyway.

eerovaher commented on astropy/utils/system_info.py:
```diff
@@ -0,0 +1,85 @@
+    return lines
+
+
+def system_info() -> int:
```
There is no need to return `0` anymore.
```suggestion
def system_info() -> None:
```

neutrinoceros replied:
done

pllim commented on astropy/utils/system_info.py:
````diff
@@ -0,0 +1,84 @@
+r"""
+A helper script to automate system info in bug reports.
+
+This can be run as
+```
+python -c "import astropy; astropy.system_info()"
+```
````
I think triple backticks is markdown, not RST. 
```suggestion
This can be run as::

    python -c "import astropy; astropy.system_info()"

```

pllim commented on astropy/utils/system_info.py:
```diff
@@ -0,0 +1,84 @@
+
+
+def system_info() -> None:
+    """
```
```suggestion
    """Print relevant system information for astropy bug reports.

```

pllim commented on astropy/utils/system_info.py:
```diff
@@ -0,0 +1,84 @@
+    Examples
+    --------
+    >>> import astropy
+    >>> astropy.system_info()
```
Hopefully this would work?
```suggestion
    >>> astropy.system_info()  # doctest: +ELLIPSIS 
```

pllim reviewed (commented):
Turns out its API doc is currently not rendered, so I think you have to add this module under "utility" in https://github.com/astropy/astropy/blob/main/docs/utils/ref_api.rst .

`@stefanv` over at Scientific Python suggested we could also parse `pyproject.toml` to gather the list of dependencies. Maybe something we can consider in the future if hardcoding is not enough.

Thanks for your patience with all the comments!

neutrinoceros commented:
I rebased and squashed all but the last two (new) commits.

pllim commented on astropy/utils/system_info.py:
```diff
@@ -0,0 +1,85 @@
+    platform.platform() = ...
+    platform.version() = ...
+    platform.python_version() = ...
+
```
Needs this so docstring does not hide the "packages" stuff below.
```suggestion
    <BLANKLINE>
```

pllim replied:
Did you miss this code suggestion?

neutrinoceros replied:
I didn't, but maybe I misinterpreted it. Is it meant to be applied as is ? If so, feel free to push to my branch, I'm signing of for the day.

pllim replied:
Yes, it is a directive for doctest to know the blank line is intentional. I will commit it. Thanks and bonne nuit!

pllim reviewed (commented):
Looks like pre-commit isn't happy about something.

I think this counts as a new feature under utils, so need a change log.

Not sure why the files diff has codecov complains. When I go into codecov page, says patch coverage is 100%. 🤷‍♀️

pllim reviewed (approved):
LGTM. Thanks!
